[REEF-1942] Add ability for .NET Client to wait for Driver to complete.

markusweimer opened the pull request:
* Added `IJobSubmissionResult.WaitForDriverToFinish()` which blocks until the Driver is done. This uses a new HTTP handler on the Java side of the Driver.
  * Added `DriverStatusHTTPHandler` to REEF / Java to query the Driver's status via HTTP and added that Handler to the Driver Configuration used by REEF.NET.
  * Added a call to block at the end of HelloREEF.

This change also adds a bunch of infrastructure to make this more informative in the future, e.g. `DriverStatus` in .NET. To keep this focused, I did not expose that machinery.

JIRA: [REEF-1942](https://issues.apache.org/jira/browse/REEF-1942)

markusweimer commented:
Whoops, forgot to make sure this passes checkstyle. Just fixed this by a force push.

motus commented on lang/cs/Org.Apache.REEF.Client/API/DriverStatus.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <summary>
+        /// Driver was RUNNING, but is no longer reachable for unknown reasons.
+        /// </summary>
+        UNKNOWN_EXITED
```
probably, `NOT_AVAILABLE`?

jwang98052 replied:
We already have ApplicationState. Driver in REEF essentially is the application. Would it cause confusion if we introduce another states.

jwang98052 replied:
Where are those status set for driver? In Java? Then should the status match the DriverStatus defined in Java?

markusweimer replied:
Hmm, that doesn't convey the status I meant. Maybe `NO_LONGER_AVAILABLE` to indicate that the status of this Driver used to be available?

markusweimer replied:
@jwang98052 said:

>We already have ApplicationState. Driver in REEF essentially is the application. Would it cause confusion if we introduce another states.

Hmm, ApplicationState is a YARN concept and distinct from the REEF concept, so it should be fine. That being said, this really should be thought through if / when we build a REEF.NET client.

markusweimer replied:
@jwang98052 said:
>Where are those status set for driver? In Java? Then should the status match the DriverStatus defined in Java?

Yes, they are set in Java, in the new class `DriverStatusHTTPHandler` introduced in this PR. That class implements `JobStatusHandler`, the interface used within the Driver to communicate the Driver (job) status to the client. It in turn relies on `ReefServiceProtos.JobStatusProto` as the representation of the Job Status. That class is generated from the protobuf spec in `lang/common/proto/reef_service_protos.proto`, which defines `State` as:

```proto
enum State {
    INIT = 0;
    RUNNING = 1;
    DONE = 2;
    SUSPEND = 3;
    FAILED = 4;
    KILLED = 5;
}
```

This is the source why I used `SUSPEND`, not `SUSPENDED`. If we don't change the proto, but want to change the code, we need logic that maps one to the other. That would introduce more potential bugs than it seems to be worth it.

motus commented on lang/cs/Org.Apache.REEF.Client/Common/JobSubmissionResult.cs:
```diff
@@ -96,6 +96,36 @@ public string GetUrlResult(string url)
+            using (StreamReader reader = new StreamReader(request.GetResponse().GetResponseStream()))
+            {
+                string statusString = reader.ReadToEnd();
+                LOGGER.Log(Level.Verbose, "Status received: " + statusString);
```
use string interpolation

markusweimer replied:
When I use the `$something{here}` syntax, VS complains. I assume we are not setup for the newest version of C#? Shall I use `string.format` instead?

motus replied:
We shall never use `string.Format()` for logging. Use `{0}` placeholders instead, e.g.
```C#
LOGGER.Log(Level.Verbose, "Status received: {0}", statusString);
```

markusweimer replied:
Thanks, fixed!

motus commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/DriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,137 @@
+
+  @Inject
+  DriverStatusHTTPHandler() {
+
```
extra blank line

markusweimer replied:
Fixed.

motus commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/DriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,137 @@
+
+  @Override
+  public void onNext(final ReefServiceProtos.JobStatusProto value) {
+    LOG.log(Level.INFO, "Received status: " + value.getState().name());
```
use string interpolation

markusweimer replied:
Done

motus commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/DriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,137 @@
+   * @param parsedHttpRequest
+   * @param response
+   * @throws IOException
+   * @throws ServletException
```
add descriptions for each parameter or remove those `@param` lines completely

markusweimer replied:
Deleted, as this is a method override anyhow.

motus commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/DriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,137 @@
+   * {@link org.apache.reef.proto.ReefServiceProtos.JobStatusProto}.
+   *
+   * @param status
+   * @return
```
either write a docstring or remove completely

markusweimer replied:
Added docstrings.

motus commented on lang/java/reef-bridge-client/src/test/java/org/apache/reef/bridge/client/TestDriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,124 @@
+
+import java.util.logging.Logger;
+
+
```
remove extra blank line

markusweimer replied:
Done.

motus commented on lang/java/reef-bridge-client/src/test/java/org/apache/reef/bridge/client/TestDriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,124 @@
+  /**
+   * An array of all statuses to test.
+   */
+  private final ReefServiceProtos.JobStatusProto[] allStatuses = new ReefServiceProtos.JobStatusProto[]{
```
add a space before `{`

markusweimer replied:
Done

motus commented on lang/java/reef-bridge-client/src/test/java/org/apache/reef/bridge/client/TestDriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,124 @@
+
+    @Override
+    public synchronized void run() {
+      synchronized (this) {
```
redundant `synchronized` block

markusweimer replied:
Great catch! Fixed.

motus commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/DriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,137 @@
+        try {
+          this.wait();
+        } catch (InterruptedException e) {
+          LOG.log(Level.FINE, "Interrupted. Ignoring.");
```
I think it is OK to throw `InterruptedException` out of the method

markusweimer replied:
I wanted to catch it and just check for messages to be sent. Is the code unsafe as is?

motus commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/DriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,137 @@
+      while (this.statusMessagesToSend.isEmpty()) {
+        try {
+          this.wait();
+        } catch (InterruptedException e) {
```
`final`

motus commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/DriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,137 @@
+    return this.lastStatus;
+  }
+
+  public String toString() {
```
add `@Override` annotation

markusweimer replied:
Done

motus commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/DriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  @Override
+  public void setUriSpecification(final String s) {
```
can use a better variable name instead of `s`

markusweimer replied:
Done.

motus commented on lang/cs/Org.Apache.REEF.Client/Common/JobSubmissionResult.cs:
```diff
@@ -96,6 +96,36 @@ public string GetUrlResult(string url)
+            {
+                string statusString = reader.ReadToEnd();
+                LOGGER.Log(Level.Verbose, "Status received: " + statusString);
+                return (DriverStatus)Enum.Parse(typeof(DriverStatus), statusString);
```
Maybe move it to the `DriverStatusMethods`? i.e. have that exact string inside the
`public static DriverStatus Parse(string statusString)`

markusweimer replied:
Good idea. Done.

motus commented:
Also, there is a unit test failure in OracleJDK8:
```text
FAILURE! - in org.apache.reef.bridge.client.TestDriverStatusHTTPHandler
testAsyncCalls(org.apache.reef.bridge.client.TestDriverStatusHTTPHandler)  Time elapsed: 0.014 sec  <<< FAILURE!
java.lang.AssertionError: expected:<INIT> but was:<null>
	at org.apache.reef.bridge.client.TestDriverStatusHTTPHandler.testAsyncCalls(TestDriverStatusHTTPHandler.java:99)
```

motus commented:
**P.S.** Just pushed a fix for the C# bug as PR #1409

jwang98052 commented on lang/cs/Org.Apache.REEF.Client/Common/JobSubmissionResult.cs:
```diff
@@ -96,6 +96,36 @@ public string GetUrlResult(string url)
+
+        private DriverStatus FetchDriverStatus()
+        {
+            string statusUrl = DriverUrl + "driverstatus/v1";
```
You must make sure the driverurl is available at the time this method is called. Or send web request only when driverUrl is available otherwise return status like UNKOWN or INIT.

markusweimer replied:
Doesn't the class make sure of that as is? `DriverUrl` merely refers to `_driverUrl`, which is set in the constructor by calling `GetDriverUrl`, which is to be implemented by every subclass. I'm assuming the subclasses to work.

If they don't, then this call will throw a `WebException`, which is documented behavior and also called out in the `Unstable` attribute. 

What else does this code need to check?

jwang98052 replied:
May be we should add sleep/retry logic in GetTrackingUrlAppId method if the file does not exist or driverUrl is not available yet. Another improvement could be to modify DriverUrl property to check if _driverUrl contains proper url, if not, recall GetDriverUrl again. It can be in separate JIRA though. 
I have created REEF-1946 for the follow up.

markusweimer replied:
Sounds good to treat this separately. Long term, I don't like this whole polling business and would much rather see a event driven approach. Let's discuss that as part of [REEF-1946]

jwang98052 commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/Constants.java:
```diff
@@ -60,6 +64,8 @@
       HttpHandlerConfiguration.CONF
           .set(HttpHandlerConfiguration.HTTP_HANDLERS, HttpServerReefEventHandler.class)
+          // Add the http status handler.
+          .set(HttpHandlerConfiguration.HTTP_HANDLERS, DriverStatusHTTPHandler.class)
```
You don't need to bind for DriverStatusHTTPHandler. This is for HttpServerReefEventHandler binding only. It is the unique one for entire server. For all the implementations of HttpHandler, REEF code will automatically detect all such handlers.

jwang98052 commented on lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/DriverStatusHTTPHandler.java:
```diff
@@ -0,0 +1,137 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.bridge.client;
```
Should this class be in common.driver.client? It is not for bridge, right?

markusweimer replied:
I write it exclusively for the bridge. I wouldn't want to promote it to the `common` package just yet. We can do that if / when we redesign the client. The Java client already has a (different) way to know when the Driver is done and doesn't need this code.

jwang98052 replied:
When you say bridge, you mean for .Net client to use, right? Our bridge code mostly is to handle the calls between .Net and Java. Although this HttpHandler is mainly used for .Net, as a web service, it can be called anywhere. From that perspective, it is not only for .Net and also it doesn't delegate any calls between Java and .Net.  That is what I mean it doesn't belong to bridge.

markusweimer replied:
I don't want to expose this as a new API to Java devs. They have a better way to do this already. I'm not set on where in the Java-for-.NET code to put it, though. Is there a better place for it?

jwang98052 commented on lang/cs/Org.Apache.REEF.Client/Common/JobSubmissionResult.cs:
```diff
@@ -96,6 +96,36 @@ public string GetUrlResult(string url)
+            }
+        }
+
+        private DriverStatus FetchDriverStatus()
```
May be we should also expose this API as public because client may want to know the driver current status. If it is only for knowing driver complete, we already have jobSubmitionResult.FinalState; that tell client the driver final state.

markusweimer replied:
As I said in the commit message, I don't want to do that just yet. A proper status API would be very useful, but needs more design considerations. For now, I want to unblock REEF-1271 without introducing any `public` APIs not needed for that goal.

jwang98052 commented:
I have done a pass. Basically it is good to control and expose our driver states through http request. The only thing is YarnRmClient already has API for the application states. Of course, the REEF driver states will be able to provide more details later and is under our control.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/API/DriverStatus.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <summary>
+        /// Driver is suspended.
+        /// </summary>
+        SUSPEND,
```
I agree that this is awkward, naming wise. I actually had `SUSPENDED` first. However, the REEF Protocols use `SUSPEND` and using any other string here would create an inconsistency with those and the Java code.

Hence, I suggest we keep this here and fix it when (if) we get around to fixing the REEF network protocols.

markusweimer commented:
@motus said
>Also, there is a unit test failure in OracleJDK8:

I cannot reproduce that on my machine on JDK8 :( Can you? Also, I am not ruling out newbie-level mistakes in the multithreaded code. Can you have a close look at both the class and the test to make sure I did not make a bubu?

markusweimer commented:
@jwang98052 said:

> You don't need to bind for DriverStatusHTTPHandler. This is for HttpServerReefEventHandler binding only. It is the unique one for entire server. For all the implementations of HttpHandler, REEF code will automatically detect all such handlers.

Do we scan the class path to make that happen? Otherwise, how does REEF find all the implementations?

(Sorry for the separate comment, but GitHub won't let me `reply` to the original one.

markusweimer commented:
@jwang98052 @motus, tanks for your comments! I have addressed them in code or comment. Please have another look.

markusweimer commented:
@motus I can now replicate the test failure. I added a loop to run the test 1k times, and it failes every once in a while. There must be a bug.

markusweimer commented:
@motus said:
> Looks good! Let's fix the bug now :)

Indeed. I should have a fix in the next couple of hours.

markusweimer commented:
@motus I fixed the bug (I believe). The test contained a race condition between unblocking the thread waiting for the new status and the thread fetching it. I fixed it by joining the waiting thread before fetching the result. in 100k repetitions on my box, this did not fail.

markusweimer commented:
This seems to have worked. All tests on Travis pass now.

jwang98052 commented:
Regarding the comments "_You don't need to bind for DriverStatusHTTPHandler. This is for HttpServerReefEventHandler binding only. It is the unique one for entire server. For all the implementations of HttpHandler, REEF code will automatically detect all such handlers."_, I removed this comment that is the reason you could not reply :). It is true for HttpHandlers in .net, where the code will auto detect all the HttpHandlers and register them. But for httpHandlers in Java, you still need to bind it to the driver configuration. That is correct.

jwang98052 commented:
Have you tested it on YARN?

markusweimer commented:
@jwang98052 said:

> Have you tested it on YARN?

Nope. Do you have an easy way to do that? I don't have an HDI cluster up :-(. Merely running HelloREEF on YARN should show us whether the new code works there.

jwang98052 commented:
Let me try to run Hello from yarn.

jwang98052 commented:
I put  LOGGER.Log(Level.Info, "WaitForDriverToFinish received: " + status); inside while loop in WaitForDriverToFinish, for locaruntime
.0.0.1:9364/.
Org.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:19:33.6697708-07:00 0001 : INFO: statusUrl: http://127.0.0.1:9364/driverstatus/v1
Org.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:19:33.7542641-07:00 0001 : INFO: Status received: INIT
Org.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:19:33.7552635-07:00 0001 : INFO: statusUrl: http://127.0.0.1:9364/driverstatus/v1
Org.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:19:33.7562639-07:00 0001 : INFO: Status received: RUNNING
Org.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:19:33.7562639-07:00 0001 : INFO: WaitForDriverToFinish received: RUNNING
Org.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:19:33.7562639-07:00 0001 : INFO: statusUrl: http://127.0.0.1:9364/driverstatus/v1
Org.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:19:46.4832204-07:00 0001 : INFO: Status received: RUNNING
Org.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:19:46.4842212-07:00 0001 : INFO: WaitForDriverToFinish received: RUNNING
Org.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:19:46.4851494-07:00 0001 : INFO: statusUrl: http://127.0.0.1:9364/driverstatus/v1
Org.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:19:47.0371590-07:00 0001 : INFO: Status received: DONE
Org.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:19:47.0371590-07:00 0001 : INFO: WaitForDriverToFinish received: DONE

In Yarn:
rg.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:33:08.3926197-07:00 0001 : INFO: statusUrl: http://10.180.113.70:2015/driverstatus/v1
rg.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:33:08.5332478-07:00 0001 : INFO: Status received: INIT
rg.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:33:08.5332478-07:00 0001 : INFO: statusUrl: http://10.180.113.70:2015/driverstatus/v1
rg.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:33:08.5332478-07:00 0001 : INFO: Status received: RUNNING
rg.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:33:08.5332478-07:00 0001 : INFO: WaitForDriverToFinish received: RUNNING
rg.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:33:08.5332478-07:00 0001 : INFO: statusUrl: http://10.180.113.70:2015/driverstatus/v1
rg.Apache.REEF.Client.Common.JobSubmissionResult Information: 0 : 2017-11-02T16:34:48.5541479-07:00 0001 : INFO: WaitForDriverToFinish received: UNKNOWN_EXITED

markusweimer commented:
Cool, so it works :)

jwang98052 commented:
I run two times on yarn, driver was successful in the end, but not sure why the state it received in the end is always UNKNOWN_EXITED. As it does receive INIT and RUNNING, means FetchDriverStatus works.

markusweimer commented:
Thanks for running it @jwang98052!

> I run two times on yarn, driver was successful in the end, but not sure why the state it received in the end is always UNKNOWN_EXITED. As it does receive INIT and RUNNING, means FetchDriverStatus works.

This state is reached because nothing prevents the Driver from shutting down before it can send `DONE` to the Client. If the Client has reached the Driver before, but can't reach it now, we know that the Driver was running and is no longer running. Which is the only thing the `public` API of this PR promises.

Of course, we should revisit this and provide more insight into the status of the Driver in the future as you suggested elsewhere on this PR. Just as soon as we finally find the time to work on a REEF.NET Client :)

jwang98052 commented:
I will merge it.
